haskell-src/Concordium/Types/Parameters.hs

```diff
@@ -27,42 +397,313 @@ import qualified Concordium.Crypto.SHA256 as Hash
+type IsAuthorizationsVersion (auv :: AuthorizationsVersion) = SingI auv
+
+-- |Witness an 'IsAuthorizationsVersion' constraint using a 'SChainParametersVersion'.
+withIsAuthorizationsVersionFor :: SChainParametersVersion cpv -> (IsAuthorizationsVersion (AuthorizationsVersionFor cpv) => a) -> a
+withIsAuthorizationsVersionFor scpv = withSingI (sAuthorizationsVersionFor scpv)
+
+-- |Witness an 'IsAuthorizationsVersion' constraint using a 'SProtocolVersion'.
+withIsAuthorizationsVersionForPV :: SProtocolVersion pv -> (IsAuthorizationsVersion (AuthorizationsVersionForPV pv) => a) -> a
+withIsAuthorizationsVersionForPV spv = withSingI (sAuthorizationsVersionForPV spv)
+
+-- |A @Conditionally b a@ is an @a@ if @b ~ 'True@, and @()@ otherwise.
+data Conditionally (b :: Bool) a where
+    CFalse :: Conditionally 'False a
+    CTrue :: !a -> Conditionally 'True a
+
+instance Functor (Conditionally b) where
+    fmap _ CFalse = CFalse
+    fmap f (CTrue v) = CTrue (f v)
+
+instance Foldable (Conditionally b) where
+    foldr _ b CFalse = b
+    foldr f b (CTrue a) = f a b
+
+    foldl _ b CFalse = b
+    foldl f b (CTrue a) = f b a
+
+    foldMap _ CFalse = mempty
+    foldMap f (CTrue a) = f a
+
+instance Traversable (Conditionally b) where
+    traverse _ CFalse = pure CFalse
+    traverse f (CTrue a) = CTrue <$> f a
+
+instance (Eq a) => Eq (Conditionally b a) where
+    CFalse == CFalse = True
+    CTrue a == CTrue b = a == b
+
+instance (Ord a) => Ord (Conditionally b a) where
+    compare CFalse CFalse = EQ
+    compare (CTrue x) (CTrue y) = compare x y
+
+instance (Show a) => Show (Conditionally b a) where
+    show CFalse = "<CFalse>"
+    show (CTrue v) = show v
+
+instance (Serialize a, SingI b) => Serialize (Conditionally b a) where
+    put CFalse = return ()
+    put (CTrue a) = put a
+
+    get = case sing @b of
+        SFalse -> pure CFalse
+        STrue -> CTrue <$> get
+
+-- |Wrap a value in a 'Conditionally' depending on the supplied 'SBool'.
+conditionally :: SBool b -> a -> Conditionally b a
+conditionally SFalse _ = CFalse
+conditionally STrue a = CTrue a
+
+-- |Perform an action conditionally on the supplied 'SBool'.
+-- This is typically used for monadic actions.
+-- The action is not performed in the 'SFalse' case.
+conditionallyA :: (Applicative f) => SBool b -> f a -> f (Conditionally b a)
+conditionallyA SFalse _ = pure CFalse
+conditionallyA STrue m = CTrue <$> m
+
+-- |A lens for accessing the contents of a 'Conditionally' when the guard is known to be 'True'.
+unconditionally :: Lens (Conditionally 'True a) (Conditionally 'True b) a b
+unconditionally f (CTrue a) = CTrue <$> f a
+
+-- |An @OParam pt cpv a@ is an @a@ if the parameter type @pt@ is supported at @cpv@, and @()@
+-- otherwise.
+data OParam (pt :: ParameterType) (cpv :: ChainParametersVersion) a where
+    NoParam :: (IsSupported pt cpv ~ 'False) => OParam pt cpv a
+    SomeParam :: (IsSupported pt cpv ~ 'True) => !a -> OParam pt cpv a
```

Comment:
Why is this a separate type, and not just an alias for Conditionally (IsSupported pt cpv). Please explain.

Reply:
I have added insightful comments for this now :-).... Thanks.